src/core/qgsabstractdatabaseproviderconnection.cpp

```diff
@@ -212,6 +213,18 @@ QList<QgsAbstractDatabaseProviderConnection::TableProperty::GeometryColumnType>
   return mGeometryColumnTypes;
 }
 
+QgsFields QgsAbstractDatabaseProviderConnection::fields( const QString &schema, const QString &tableName ) const
+{
+  QgsVectorLayer vl { tableUri( schema, tableName ), QStringLiteral( "temp_layer" ), mProviderKey };
```

Comment:
If you specify LayerOptions here you could avoid some unnecessary work -- e.g. skipCrsValidation should be true, and maybe setting readExtentFromXml to true would also avoid some unneeded calculations